compositor_web/src/wasm/input_uploader.rs

```diff
@@ -1,37 +1,68 @@
 use std::{collections::HashMap, sync::Arc, time::Duration};
 
 use compositor_render::{Frame, FrameData, FrameSet, InputId};
-use wasm_bindgen::JsValue;
+use futures::future::join_all;
+use js_sys::Object;
+use tracing::error;
+use wasm_bindgen::{JsCast, JsValue};
+use wasm_bindgen_futures::JsFuture;
+use web_sys::{VideoFrameCopyToOptions, WorkerGlobalScope};
 
-use super::types;
+use super::types::{self, ObjectExt};
 
-#[derive(Default)]
 pub struct InputUploader {
     textures: HashMap<InputId, Arc<wgpu::Texture>>,
+
+    /// On most platforms it's more performant to copy VideoFrame data to CPU
+    /// and then upload it wgpu::Texture. But on macOS using dedicated wgpu copy_external_image_to_texture function
+    /// results in better performance.
+    use_copy_external: bool,
+}
+
+impl Default for InputUploader {
+    fn default() -> Self {
+        let user_agent = match web_sys::window() {
```

Comment:
for simplicity I would do that in js and pass to WASM selected mode when renderer is instantiated